src/playground-tab-bar.ts

```diff
@@ -0,0 +1,376 @@
+    if (!controls) {
+      return;
+    }
+    controls.state = 'newfile';
+    controls.anchorElement = event.target as HTMLElement;
+  }
+
+  private _onNewFile(event: CustomEvent<{filename: string}>) {
+    this._activeFileName = event.detail.filename;
+    // TODO(aomarks) We should focus the editor here. However,
+    // CodeMirror.focus() isn't working for some reason.
+  }
+
+  /**
+   * Whenever a file is created, deleted, or renamed, figure out what the best
+   * new active tab should be.
+   */
+  private _setNewActiveFile() {
+    if (this._files.length === 0) {
+      this._activeFileIndex = 0;
+      this._activeFileName = '';
+      return;
+    }
+
+    // Stay on the same filename if it's still around, even though its index
+    // might have changed.
+    if (this._activeFileName) {
+      const index = this._files.findIndex(
+        (file) => file.name === this._activeFileName
+      );
+      if (index >= 0) {
+        this._activeFileIndex = index;
+        return;
+      }
+    }
+
+    // The active file was deleted, and it was the final tab. Move to the next
+    // tab down.
+    if (this._activeFileIndex > this._files.length - 1) {
+      this._activeFileIndex = this._files.length - 1;
+      this._activeFileName = this._files[this._activeFileIndex].name;
+      return;
+    }
+
+    // The active file was deleted, and another file has now taken the tab
+    // position of the old one. Select that file.
+    this._activeFileName = this._files[this._activeFileIndex].name;
+  }
+}
+
+/**
+ * Internal element for tabs within <playground-tab-bar>.
+ *
+ * This is a subclass of <mwc-tab>. We subclass because <mwc-tab> only supports
+ * text labels, and has no way to slot in our menu button.
+ *
+ * Note we can't subclass TabBase, because <mwc-tab-bar> relies on an instanceof
+ * check for Tab.
+ */
+@customElement('playground-tab')
+class PlaygroundTab extends Tab {
+  /**
+   * Whether to show the 3-dots menu button.
+   */
+  @property({type: Boolean})
+  showMenuButton = false;
+
+  /**
+   * 0-indexed position of this tab in the tab list.
+   *
+   * Note this could be parsed from the 1-indexed `id` that <mwc-tab-bar> sets
+   * on each instance after slotting, but taking a property here is simpler.
+   */
+  index = 0;
+
+  static styles = ([
+    mwcTabStyle,
+    css`
+      :host {
+        /* Vertically center the menu button. */
+        display: flex;
+        align-items: center;
+      }
+
+      .mdc-tab {
+        /* A little extra room for the menu button. */
+        padding-right: calc(var(--mdc-tab-horizontal-padding, 24px) + 8px);
+      }
+
+      .menu-button {
+        /* Shift the menu button to be inside the tab itself. */
+        margin-left: -24px;
+        z-index: 1;
+        opacity: 70%;
+        --mdc-icon-button-size: 24px;
+        --mdc-icon-size: 16px;
+      }
+
+      .menu-button:hover {
+        opacity: 100%;
+      }
+    `,
+  ] as unknown) as CSSResult;
```

Comment:
argh, I'm glad we fixed this in lit-next

Reply:
Acknowledged.